src/store.rs

```diff
@@ -471,335 +587,287 @@ impl<'a> Run<'a> {
+                error!(
+                    "Fatal: failed to get position in file {}: {}",
+                    tmp_path.display(), err
+                );
+                return Err(UpdateError::Fatal)
+            }
+        };
 
-    /// A batch of the changed objects.
-    objects: sled::Batch,
-}
+        loop {
+            match objects() {
+                Ok(Some(object)) => {
+                    if let Err(err) = object.write(&mut tmp_file) {
+                        error!(
+                            "Fatal: failed to write to file {}: {}",
+                            tmp_path.display(), err
+                        );
+                        return Err(UpdateError::Fatal)
+                    }
+                }
+                Ok(None) => break,
+                Err(err) => {
+                    drop(tmp_file);
+                    if let Err(err) = fs::remove_file(&tmp_path) {
+                        error!(
+                            "Fatal: failed to delete file {}: {}",
+                            tmp_path.display(), err
+                        );
+                        return Err(UpdateError::Fatal)
+                    }
+                    return Err(err)
+                }
+            }
+        }
 
-impl<'a> RepositoryUpdate<'a> {
-    /// Creates an update from the manifest URI and the transactions.
-    fn new(
-        manifest: &'a uri::Rsync,
-    ) -> Self {
-        RepositoryUpdate {
-            manifest,
-            new_manifest: None,
-            objects: sled::Batch::default(),
+        drop(tmp_file);
+        let existing = self.file.is_some();
+        drop(self.file.take());
+
+        if existing {
+            if let Err(err) = fs::remove_file(&self.path) {
+                error!(
+                    "Fatal: cannot delete file {}: {}",
+                    self.path.display(), err
+                );
+                return Err(UpdateError::Fatal)
+            }
+        }
+        else if let Some(path) = self.path.parent() {
+            if let Err(err) = fs::create_dir_all(&path) {
+                error!(
+                    "Fatal: cannot create direcory {}: {}",
+                    path.display(), err
+                );
+                return Err(UpdateError::Fatal)
+            }
+        }
+        if let Err(err) = fs::rename(&tmp_path, &self.path) {
+            error!(
+                "Fatal: cannot move {} to {}: {}",
+                tmp_path.display(), self.path.display(), err
+            );
+            return Err(UpdateError::Fatal)
+        }
+        let mut file = match File::open(&self.path) {
+            Ok(some) => some,
+            Err(err) => {
+                error!(
+                    "Fatal: failed to open {}: {}",
+                    self.path.display(), err
+                );
+                return Err(UpdateError::Fatal)
+            }
+        };
+        if let Err(err) = file.seek(SeekFrom::Start(tmp_object_start)) {
```

Comment:
In what situation would this actually error, if the `rename` and `open` were successful?

Reply:
Who knows what the file systems does. The method can error, so we need to handle it.